File: Modeling .NET/VB/Bank/Bank/BankSimulation.vb
Imports Simulation

''' <summary>
''' Класс BankSimulation имитирует работу банка
''' </summary>
''' <remarks></remarks>
Public Class BankSimulation
    Inherits SimProc

    ''' <summary>
    ''' Статистика по загруженности кассира
    ''' </summary>
    ''' <remarks></remarks>
    Public CashStat As ServiceStatistics

    ''' <summary>
    ''' Гистограмма по времени нахождения клиентов в банке
    ''' </summary>
    ''' <remarks></remarks>
    Public InBankHist As Histogram

    ''' <summary>
    ''' Статистика по времени нахождения клиентов в банке
    ''' </summary>
    ''' <remarks></remarks>
    Public InBankStat As Statistics

    ''' <summary>
    ''' Количество клиентов, обслуженных без ожидания
    ''' </summary>
    ''' <remarks></remarks>
    Public NotWaited As Integer

    ''' <summary>
    ''' Очередь клиентов, ожидающих обслуживания
    ''' </summary>
    ''' <remarks></remarks>
    Public Queue As List

    ''' <summary>
    ''' Кассир
    ''' </summary>
    ''' <remarks></remarks>
    Friend Cash As Cashman

    ''' <summary>
    ''' Генератор клиентов
    ''' </summary>
    ''' <remarks></remarks>
    Friend Generator As ClientGenerator

    ''' <summary>
    ''' Генератор случайных чисел, управляющий работой кассира
    ''' </summary>
    ''' <remarks></remarks>
    Friend Shared RandCashman As Simulation.Random

    ''' <summary>
    ''' Генератор случайных чисел, управляющий прибытием клиентов
    ''' </summary>
    ''' <remarks></remarks>
    Friend Shared RandClient As Simulation.Random

    ''' <summary>
    ''' основной алгоритм работы имитации
    ''' </summary>
    ''' <remarks></remarks>
    Protected Overrides Sub Execute()
        Generator.Activate()
        Passivate()
    End Sub

    ''' <summary>
    ''' Создание инфраструктуры имитации
    ''' </summary>
    ''' <remarks></remarks>
    Protected Overrides Sub Init()
        MyBase.Init()
        Cash = New Cashman
        Generator = New ClientGenerator
        Queue = New List("Очередь ожидания")
        InBankStat = New Statistics("Время нахождения клиентов в системе")
        InBankHist = New Histogram(Params.HistMin, Params.HistStep, Params.HistStepCount, _
                                   "Время нахождения клиентов в системе")
        CashStat = New ServiceStatistics("Занятость кассира")
        NotWaited = 0
    End Sub

    ''' <summary>
    ''' Завершение работы имитации
    ''' </summary>
    ''' <remarks></remarks>
    Public Overrides Sub Finish()
        Queue.Finish()
        Cash.Finish()
        Generator.Finish()
        MyBase.Finish()
    End Sub

    ''' <summary>
    ''' Коррекция статистики к текущему времени
    ''' </summary>
    ''' <remarks></remarks>
    Public Overrides Sub StopStat()
        MyBase.StopStat()
        CashStat.StopStat()
        Queue.StopStat()
    End Sub
End Class

File: Modeling .NET/CSharp/3.1.5 BankVisualNoProcs/BankVisual/BankSimulation.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Simulation;

namespace BankVisual
{
    using SimRandom = Simulation.Random;

    /// <summary>
    /// Имитация обслуживания клиентов в банке
    /// </summary>
    public class BankSimulation : SimProc
    {
        /// <summary>
        /// Точечная гистограмма по времени пребывания в банке
        /// </summary>
        public Histogram InBankHist;

        /// <summary>
        /// Количество клиентов, обслуженных без ожидания в очереди
        /// </summary>
        public int NotWaited;

        /// <summary>
        /// Очередь ожидания
        /// </summary>
        public List Queue;

        /// <summary>
        /// Статистика по занятости кассира
        /// </summary>
        public ServiceStatistics CashStat;

        /// <summary>
        /// Генератор случайных чисел, управляющий работой клиентов
        /// </summary>
        public static SimRandom RandClient;

        /// <summary>
        /// Генератор случайных чисел, управляющий работой кассира
        /// </summary>
        public static SimRandom RandCashman;

        /// <summary>
        /// Точечная статистика по времени пребывания в банке
        /// </summary>
        public Statistics InBankStat;

        /// <summary>
        /// Кассир
        /// </summary>
        internal Cashman Cash;

        /// <summary>
        /// Генератор клиентов
        /// </summary>
        internal ClientGenerator Generator;

        /// <summary>
        /// Алгоритм имитации
        /// </summary>
        protected override void  Execute()
        {
            // Запустить генератор клиентов
            Generator.Activate();
            // Ждать окончания обслуживания
            Passivate();
        }

        /// <summary>
        /// Создание объектов имитации
        /// </summary>
        protected override void Init()
        {
            base.Init();
            Cash = new Cashman();
            Generator = new ClientGenerator();
            Queue = new List("Очередь ожидания");
            InBankStat = new Statistics("Время нахождения клиентов в системе");
            InBankHist = new Histogram(Param.HistMin, Param.HistStep, Param.HistStepCount, "Время нахождения клиентов в системе");
            CashStat = new ServiceStatistics("Занятость кассира");
            NotWaited = 0;
        }

        /// <summary>
        /// Удаление имитации
        /// </summary>
        public override void Finish()
        {
            // Требуется удалить все процессы
            Queue.Finish();
            Cash.Finish();
            Generator.Finish();
            // Остальные объекты имитации удалять не требуется
            base.Finish();
        }

        /// <summary>
        /// Коррекция статистики
        /// </summary>
        public override void StopStat()
        {
            base.StopStat();
            CashStat.StopStat();
            Queue.StopStat();
        }
    }
}
